# Questão 1

In [165]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import math

# Lista de Cáculo Numérico

Problema 1: Considere o seguinte sistema de equações lineares:
 $$
 \left\{\begin{matrix}
x_{1} + 2x_{2} - x_{3} + 0x_{4} = 1 &  &  & \\ 
2x_{1} - x_{2} + 0x_{3} + 0x_{4} = 1 &  &  & \\ 
0x_{1} - x_{2} + 2x_{3} - x_{4} = 1 &  &  & \\ 
0x_{1} + 0x_{2}-x_{3}+2x_{4} = 1 &  &  & 
\end{matrix}\right.
 $$

## a: Mostre que esse sistema não satisfaz o critério de linhas

O critério de linhas diz que o módulo de cada elemento da diagonal principal, deve ser maior que a soma dos módulos dos elementos restantes da linha.

Ex:
    $$
        \left \| a_{11} \right \| > \left \| a_{12} \right \| +\left \| a_{13} \right \|
    $$

O critério das linhas pode ser descrito da seguinte forma:

$$
\alpha _{k} = \frac{1}{\left \| a_{kk} \right \|}\sum_{\underset{j\neq k}{j=1}}^{n}\left \| a_{kj} \right \|
$$

Então será feito o seguinte: Será determinada uma função que receberá um array np, e tratará conforme a definição acima. Caso o coeficiente máximo encontrado seja menor que 1, verifica-se a convergência.

In [166]:
def jacobi(arr):
    termos = []
    coeficientes = 0
    arr = np.absolute(arr)
    
    for i in range(arr.shape[0]):
        for j in range(arr.shape[0]):
            if(j != i):
                coeficientes += arr[i][j]
            
        termos.append(coeficientes * (1/arr[i][i]))
        coeficientes=0
            
    return termos
        
    

In [167]:
arr = np.array([[1,2,-1,0],[2,-1,0,0],[0,-1,2,-1],[0,0,-1,2]])
np.max(jacobi(arr))

3.0

Percebemos, pelo algoritmo, que o sistema não obedece o critérios de linhas pois o coeficiente máximo alcançado é 3, logo, não converge para Jacobi.

## b: Enuncie o critério de Sassenfeld e mostre que esse sistema não satisfaz o critério de Sassenfeld

O critério de Sassenfeld é muito parecido com o critério das linhas, porém, é multiplicado o coeficiente anterior pelo número adjacente ao diagonal anterior. O algorítmo será desenvolvido a partir da definição algébrica.

O critério de Sessenfeld pode ser descrito algebricamente da seguinte forma:

$$
\beta _{k} = \frac{1}{a_{kk}}\left ( \sum_{j=1}^{k-1}\left \| a_{kj} \right \| \beta _{j}+\sum_{j=k+1}^{n}\left \| a_{kj} \right \| \right )
$$

In [168]:
def sassenfeld(arr):
    termos = []
    coeficientes = 0
    arr = np.absolute(arr)
    print(arr)
    
    for i in range(arr.shape[0]):
        coeficientes = 0
        if(i>1):
            for j in range(i):
                coeficientes += arr[i][j] * termos[j]
            for j in range(i+1, arr.shape[0]):
                coeficientes += arr[i][j]
        elif(i == 1):
            coeficientes += arr[i][0] * termos[0]
            for j in range(i+1, arr.shape[0]):
                coeficientes += arr[i][j]
        elif(i==0):
            for j in range(i+1, arr.shape[0]):
                coeficientes += arr[i][j]
        termos.append(coeficientes * (1/arr[i][i]))
            
            
    return termos
   

In [169]:
arr = np.array([[1,2,-1,0],[2,-1,0,0],[0,-1,2,-1],[0,0,-1,2]])
sassenfeld(arr)

[[1 2 1 0]
 [2 1 0 0]
 [0 1 2 1]
 [0 0 1 2]]


[3.0, 6.0, 3.5, 1.75]

Percebe-se que o sistema, em sua forma atual, não satisfaz o critério de Sassenfeld pois o máximo é 6, que é maior que 1

## c: O que se pode afirmar sobre a convergência dos métodos de Gauss-Jacobi e Gauss-seidel, quando aplicados a este sistema?

Pode-se afirmar que o sistema não satisfaz nenhum dos critérios

## d: Mostre que o sistema obtido permutando-se as duas primeiras equações satisfaz o critério de Sassenfeld

In [170]:
arr = np.array([[2,-1,0,0],[1,2,-1,0],[0,-1,2,-1],[0,0,-1,2]])
sassenfeld(arr)

[[2 1 0 0]
 [1 2 1 0]
 [0 1 2 1]
 [0 0 1 2]]


[0.5, 0.75, 0.875, 0.4375]

Realmente satisfaz, pois o máximo é 0.875, que é menor que 1

## e: Usando o método de Gauss-Seidel, determine a solução aproximada do sistema, com a permutação sugerida no item anterior e erro = 0.001

Teremos que fazer uma função que recebe nosso array np, para que possa ser aplicado o método de Gayuss-Seidel.

O método de Gauss-Seidel pode ser escrito algebricamente da seguinte forma:

$$
x^{k+1} = Tx^{k} + C
$$
onde
$$
T = -L^{-1}_{*}U
$$
e
$$
C = L^{-1}_{*}b
$$

Sabemos que $L_{*}$ é a matriz triangular inferior com a diagonal da matriz principal, e $U$ é a matriz triangular superior com a diagonal zerada.

In [171]:
def gaussSeidel(arr,x,b):
    
    
    #pegando matriz triangular inferior
    l = np.copy(np.tril(arr))
    
    #pegando matriz triangular superior com diagonal zerada
    r = np.copy(np.triu(arr))
    np.fill_diagonal(r, 0)
    
    
    #t corresponde a primeira parte da equação de gauss-seidel
    t = np.copy(np.matmul(-np.linalg.inv(l),r))
    
    #c corresponde a segunda parte da equação de gauss-seidel
    c = np.copy(np.matmul(np.linalg.inv(l),b))
    
    
    x = np.matmul(t,x) + c
    
    return x


In [172]:
arr = np.array([[2,-1,0,0],[1,2,-1,0],[0,-1,2,-1],[0,0,-1,2]])
x = np.array([[0.5],[0.5],[0.5],[0.5]])
b = np.array([[1],[1],[1],[1]])

for i in range(20):
    x = gaussSeidel(arr,x,b)
print(x)

result = x[0] + 2*x[1] - x[3]
print('')
print(result)

[[0.90909089]
 [0.8181818 ]
 [1.54545452]
 [1.27272726]]

[1.27272723]
